src/interface/StateDescriptor.hpp

```diff
@@ -55,23 +55,32 @@ class StateDescriptor {
     // field addition / retrieval routines
     // add a field with associated metadata
     bool AddField(const std::string& field_name, Metadata& m, DerivedOwnership owner=DerivedOwnership::unique) {
-      const std::string& assoc = m.getAssociated();
-      if (!assoc.length()) m.Associate(field_name);
-      auto miter = _metadataMap.find(field_name);
-      if (miter != _metadataMap.end()) { // this field has already been added
-        Metadata& mprev = miter->second;
-        if (owner == DerivedOwnership::unique) {
-          throw std::invalid_argument("Field "+field_name+" add with DerivedOwnership::unique already exists");
+      if (m.IsSet(Metadata::Sparse)) {
+        auto miter = _sparseMetadataMap.find(field_name);
+        if (miter != _sparseMetadataMap.end()) {
+          miter->second.push_back(m);
+        } else {
+          _sparseMetadataMap[field_name] = {m};
         }
-        if (mprev != m) {
-          throw std::invalid_argument("Field " + field_name + " already exists with different metadata");
-        }
-        return false;
       } else {
-        _metadataMap[field_name] = m;
-        m.Associate("");
-        return true;
+        const std::string& assoc = m.getAssociated();
+        if (!assoc.length()) m.Associate(field_name);
+        auto miter = _metadataMap.find(field_name);
+        if (miter != _metadataMap.end()) { // this field has already been added
+          Metadata& mprev = miter->second;
+          if (owner == DerivedOwnership::unique) {
+            throw std::invalid_argument("Field "+field_name+" add with DerivedOwnership::unique already exists");
+          }
+          if (mprev != m) {
+            throw std::invalid_argument("Field " + field_name + " already exists with different metadata");
+          }
+          return false;
+        } else {
+          _metadataMap[field_name] = m;
+          m.Associate("");
+        }
       }
+      return true;
```

Comment:
The `StateDescriptor` could probably use some documentation explaining its purpose... e.g., that it is primarily used to tell the mesh about the requirements of an individual package, e.g., it tells the meshblocks how to initialize storage.

Reply:
done.